### PR TITLE
Disallow the use of quantity input for 'decimalyear' format.

### LINKED ISSUE #14541: Problem with "decimalyear" applied to MaskedColumn type
### Description

The "decimalyear" type produces odd errors when applied to the MaskedColumn type.
The "jyear" type does not, and appears to behave properly.

### Expected behavior

I would expect "decimalyear" and "jyear" to work similarly, although the difference between them is not clear from the documentation.

### How to Reproduce
```python
# Begin decimalyear_test.py
# A simple illustration of possible bug in "decimalyear"

import astropy
from astropy.coordinates import SkyCoord
from astropy.units import Quantity
from astroquery.gaia import Gaia
from astropy.time import Time

coord = SkyCoord(ra=0.0, dec=0.0, unit=astropy.units.deg)
width = Quantity(0.05, unit=astropy.units.deg)
height = Quantity(0.05, unit=astropy.units.deg)

Gaia.MAIN_GAIA_TABLE = "gaiadr3.gaia_source" 
searchout  = Gaia.query_object_async(coordinate=coord, width=width, height=height)	

epochs=searchout['ref_epoch']
print(epochs,"\n")

print("epochs is instance of MaskedColumn:", isinstance(epochs, astropy.table.column.MaskedColumn),"\n")

print("epochs in jyear: ",Time(epochs,format='jyear'),"\n")
print("epochs in decimalyear: ",Time(epochs,format='decimalyear'))

print("\n")
epoch2=Time(2016.0,format='jyear')
print("epoch2 in jyear=", epoch2)

epoch3=Time(2016.0,format='decimalyear')
print("epoch3 in decimalyear=", epoch3)

# End decimalyear_test.py
```
### Versions

Linux-5.19.0-35-generic-x86_64-with-glibc2.35
Python 3.9.13 (main, Aug 25 2022, 23:26:10) 
[GCC 11.2.0]
astropy 5.1

### PR DESCRIPTION
**Pull request by mhvk**
Using quantities with units of time for ``Time`` format 'decimalyear' will now raise an error instead of converting the quantity to days and then interpreting the value as years. An error is raised instead of attempting to interpret the unit as years, since the interpretation is ambiguous: in 'decimaltime' years are equal to 365 or 366 days, while for regular time units the year is defined as 365.25 days.

Fixes #14541

### CODE REVIEW

**Comment by mhvk** [comment]
dev failures unrelated (URL timeouts)

**Comment by mhvk** [comment]
I think it is temporary - I can open the URL now.

**Comment by pllim** [comment]
I restarted the remote data job

**Comment by pllim** [comment]
There are two real failures on the remote data job (that also picks up numpy dev), so I cannot tell if they are related to your PR or not: https://github.com/astropy/astropy/actions/runs/4495274221/jobs/7921203286?pr=14566

**Comment by mhvk** [comment]
Both failures are weird, but definitely unrelated to this PR. It looks like a location has moved according to google (I guess we just pick a more guaranteed example...), while something much have changed with `np.fix`. Will investigate...

File: astropy/time/formats.py [hunk 1]
```diff
@@ -584,6 +584,16 @@ class TimeDecimalYear(TimeNumeric):
 
     name = "decimalyear"
 
+    def _check_val_type(self, val1, val2):
+        if hasattr(val1, "to") and hasattr(val1, "unit") and val1.unit is not None:
```
Comment (eerovaher) [inline astropy/time/formats.py:588]:
```suggestion
        if hasattr(val1, "to") and getattr(val1, "unit", None) is not None:
```

Comment (mhvk) [reply astropy/time/formats.py:588]: Yes, that is nicer. I should then change it for `byear` as well.

File: astropy/time/formats.py [hunk 2]
```diff
@@ -584,6 +584,16 @@ class TimeDecimalYear(TimeNumeric):
...
+            raise ValueError(
```
[the remark this answers is not in the document]
Comment (mhvk) [reply astropy/time/formats.py:589]: I feel that would indeed be more logical, but we don't do it for `byear` and I'm not sure if I should change that one... Let me ping @taldcroft

**Comment by taldcroft** [comment]
Looking at the code it looks like there is another related mistake that the `jyear` (`TimeJulianEpoch`) format does not check for a `Quantity` input. It seems like the `_check_val_type(self, val1, val2):` method in `TimeBesselianEpoch` should instead be in `TimeEpochDate`.

But with 3 classes using the same checker maybe it is time for a mixin class that provides just that `_check_val_type` method?

**Comment by mhvk** [comment]
@taldcroft - `jyear` can work with `Quantity`, because it has the same definition of a year (of 365.25 days), so I think it is just `byear` and `decimalyear`.
```
In [3]: Time(365.25*2000*u.day, format='jyear')
Out[3]: <Time object: scale='tt' format='jyear' value=2000.0>
```

But I'm happy to combine `byear` and `decimalyear`. What do you think about the comment of @eerovaher that it should be a `TypeError` instead of a `ValueError`?

**Comment by taldcroft** [comment]
> jyear can work with Quantity, because it has the same definition of a year (of 365.25 days), so I think it is just byear and decimalyear.

Personally I find this surprising, or at least contrary to the documentation for the float `TimeEpoch` classes. E.g. `"""Julian Epoch year as floating point value(s) like 2000.0."""`. For me, all the Time Format classes are representations of a point in time and cannot be set with a `Quantity` which is an interval of time. I would consistently apply that even for time formats that are defined as an interval of time since an epoch.

Alas that ship has already sailed and some of the epoch classes will accept a Quantity and some won't, so I agree with the approach here to fix `DecimalYear`. For extra credit you might expand the PR scope a bit and update the doc strings to indicate which ones also accept a `Quantity`.

File: astropy/time/formats.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mhvk) [reply astropy/time/formats.py:589]: OK, let's leave the type of error for another PR

**Comment by mhvk** [comment]
@taldcroft - I added an extra commit that refactors out a new `TimeNumericNoQuantity` class and uses that for both `byear` and `decimalyear`. I also adjusted a few docstrings, but did not make a note for all. I considered editing the time documentation (perhaps the description of [`val`](https://docs.astropy.org/en/latest/time/index.html#val)) but then realized that this part also mentions that `scale` is a required parameter, so it would be more of an update than I have time for right now...

File: astropy/time/formats.py (hunk 1 above)
Comment (mhvk) [reply astropy/time/formats.py]: Implemented.

**Comment by mhvk** [comment]
OK, tests all passed, ready for final review...

**Comment by mhvk** [comment]
Thinking about this again, I'm not 100% sure I like the new `TimeNumericNoQuantity` class as it introduces API that I'm not sure we want to support forever. Shall I just revert that part, duplicating the check? (But keeping the docstring updates.)

**Comment by taldcroft** [comment]
@mhvk - one possibility is a helper function or static method that implements the no-Quantity check. Lately I find myself trying to avoid so much inheritance. An advantage of helper functions is that they can be easier to test in unit tests, and it is definitely easier to understand when reading the code. E.g.
```
class TimeDecimalYear(TimeNumeric):
    def _check_val_type(self, val1, val2):
        _check_val_type_not_quantity(self.name, val1, val2)  # check val1 and val2 are not Quantity
        return super()._check_val_type(val1, val2)
```

**Comment by mhvk** [comment]
OK, I went with your suggestion - that does leave us free to change things if we find a better way to do this. (E.g., I experimented with `unit = None`, which worked except it is also used to determine the number of digits in string output. Which is probably better done another time!

**Review by taldcroft — APPROVED** [review]
@mhvk - looks great! CI hasn't fully passed yet, but assuming that it does you can click the "CI passed" checkbox and merge.

**Comment by pllim** [comment]
Is there a backport to v5.2.x ? I don't see it.

**Pull request merged**